capsules/extra/src/net/thread/driver.rs

```diff
@@ -0,0 +1,744 @@
+    net_cap: &'static NetworkCapability,
+
+    /// Frame counter for Thread MLE
+    frame_count: Cell<u32>,
+
+    /// Stored Thread network containing mac/MLE key
+    networkkey: MapCell<NetworkKey>,
+
+    /// Length of the message passed to the crypto engine
+    crypto_sizelock: MapCell<usize>,
+}
+
+// Note: For now, we initialize the Thread state as empty.
+// We replace the Thread state when the first userspace
+// application calls the Thread capsule to initiate a Thread network.
+// This serves to "lock" the Thread capsule to only one application.
+// For now, Tock only supports one application using the Thread network.
+// After the network is "locked" to one application, other userspace
+// applications requesting to join a Thread network will fail.
+impl<'a, A: time::Alarm<'a>> ThreadNetworkDriver<'a, A> {
+    pub fn new(
+        sender: &'a dyn UDPSender<'a>,
+        aes_crypto: &'a dyn AES128CCM<'a>,
+        alarm: &'a A,
+        grant: Grant<App, UpcallCount<1>, AllowRoCount<{ ro_allow::COUNT }>, AllowRwCount<0>>,
+        src_mac_addr: [u8; 8],
+        max_tx_pyld_len: usize,
+        port_table: &'static UdpPortManager,
+        send_buffer: SubSliceMut<'static, u8>,
+        recv_buffer: SubSliceMut<'static, u8>,
+        driver_send_cap: &'static dyn UdpDriverCapability,
+        net_cap: &'static NetworkCapability,
+    ) -> ThreadNetworkDriver<'a, A> {
+        ThreadNetworkDriver {
+            sender: sender,
+            aes_crypto: aes_crypto,
+            alarm: alarm,
+            apps: grant,
+            src_mac_addr: src_mac_addr,
+            max_tx_pyld_len: max_tx_pyld_len,
+            port_table: port_table,
+            send_buffer: MapCell::new(send_buffer),
+            recv_buffer: MapCell::new(recv_buffer),
+            state: MapCell::empty(),
+            driver_send_cap: driver_send_cap,
+            net_cap: net_cap,
+            frame_count: Cell::new(5),
+            networkkey: MapCell::empty(),
+            crypto_sizelock: MapCell::empty(),
+        }
+    }
+
+    /// Takes the MLE and MAC keys and replaces the networkkey
+    pub fn set_networkkey(&self, mle_key: [u8; 16], mac_key: [u8; 16]) {
+        self.networkkey.replace(NetworkKey { mle_key, mac_key });
+    }
+
+    /// Thread initialization function to bing Thread to the default Thread MLE
+    /// ports
+    pub fn init_thread_binding(&self) -> (UdpPortBindingTx, UdpPortBindingRx) {
+        // Initialization function to bind thread on the Thread UDP port needed to send MLE
+        match self.port_table.create_socket() {
+            Ok(socket) => match self
+                .port_table
+                .bind(socket, THREAD_PORT_NUMBER, self.net_cap)
+            {
+                Ok((udp_tx, udp_rx)) => (udp_tx, udp_rx),
+                Err(sock) => panic!("failed binding of port to socket - {:?}", sock),
+            },
+            Err(return_code) => panic!("Error in retrieving socket - {:?}", return_code),
```

Comment:
Again, do we think these are quite plausible runtime issues, or only triggered if the implementation itself is bad?

Generally, my thought is that we shouldn't be calling `panic` from a capsule for runtime issues. But a `panic` that would only occur if configuration is wrong or if there's an implementation bug, seems reasonable.

Reply:
I agree with your thinking about only calling `panic` on configuration issues. This case you cited here interestingly spans both runtime / configuration. Thread will be unable to operate if it cannot bind to port 19788 for UDP (unable to send MLE packets).   `init_thread_binding()` is called in the Thread component and binds Thread to the needed UDP port. This is predominantly a configuration case (since it is occurring in components). Rethinking this now,  we can move this functionality to entirely exist in the component. 

This brings up an interesting design decision we should discuss in the next networking wg. 

> Should Thread be the only entity capable of binding to 19788?
_(I lean away from this because this seems needlessly restrictive and we can bind Thread to this port once an application requests to create or join a Thread network)_

In the short term and for this pull request, I think this binding should be moved to the component and the error checking removed. The UDP send request will fail if this binding does not occur, and this case is handled in the Thread implementation (terminates the Thread joining).

Reply:
Correction: UDP `driver_send_to(...)` does not require the port to be specified to send udp packets. This means Thread will not be able to receive packets and silently fail/eventually time out if this binding somehow fails. Once again, I think this is fine for now, but long term will need a more robust error handling.